profiling/src/profiling.rs

```diff
@@ -0,0 +1,806 @@
+                Ok(status) => {
+                    if status >= 400 {
+                        warn!(
+                            "Unexpected HTTP status when sending profile (HTTP {}).",
+                            status
+                        )
+                    } else {
+                        info!("Successfully uploaded profile (HTTP {}).", status)
+                    }
+                }
+                Err(err) => {
+                    warn!("Failed to upload profile: {}", err)
+                }
+            });
+    }
+}
+
+impl Profiler {
+    pub fn new() -> Self {
+        let (message_sender, message_receiver) = crossbeam_channel::bounded(100);
+        let (time_collector_cancel_sender, time_collector_cancel_receiver) =
+            crossbeam_channel::bounded(0);
+        let vm_interrupt_lock = Arc::new(Mutex::new(HashSet::with_capacity(1)));
+        let (upload_sender, upload_receiver) = crossbeam_channel::bounded(UPLOAD_CHANNEL_CAPACITY);
+        let time_collector = TimeCollector {
+            message_receiver,
+            cancel_receiver: time_collector_cancel_receiver,
+            vm_interrupt_lock: vm_interrupt_lock.clone(),
+            upload_sender,
+            wall_time_period: WALL_TIME_PERIOD,
+            upload_period: UPLOAD_PERIOD,
+        };
+        let uploader = Uploader { upload_receiver };
+        Profiler {
+            message_sender,
+            time_collector_cancel_sender,
+            vm_interrupt_lock,
+            time_collector_handle: std::thread::spawn(move || time_collector.run()),
+            uploader_handle: std::thread::spawn(move || uploader.run()),
+        }
+    }
+
+    pub fn add_interrupt(
+        &self,
+        engine_ptr: *const AtomicBool,
+        interrupt_count_ptr: *const AtomicU32,
+    ) {
+        match self.vm_interrupt_lock.lock() {
+            Ok(mut interrupts) => {
+                interrupts.insert(VmInterrupt {
+                    interrupt_count_ptr,
+                    engine_ptr,
+                });
+            }
+            Err(err) => {
+                error!("failed to acquire vm_interrupt lock: {}", err);
+            }
+        };
+    }
+
+    pub fn remove_interrupt(
+        &self,
+        engine_ptr: *const AtomicBool,
+        interrupt_count_ptr: *const AtomicU32,
+    ) {
+        match self.vm_interrupt_lock.lock() {
+            Ok(mut interrupts) => {
+                interrupts.remove(&VmInterrupt {
+                    interrupt_count_ptr,
+                    engine_ptr,
+                });
+            }
+            Err(err) => {
+                error!("failed to acquire vm_interrupt lock: {}", err);
+            }
+        };
+    }
+
+    pub fn send_sample(&self, message: SampleMessage) -> Result<(), TrySendError<SampleMessage>> {
+        self.message_sender.try_send(message)
+    }
+
+    pub fn stop(self) {
+        // todo: what should be done when a thread panics?
```

Comment:
Intentionally left TODO?

Reply:
Yes, but also I really don't know what ought to be done. In general, I don't think anything can be done except to propagate the panic.

Reply:
FWIW this is a Hard™ problem, and we build Mork with `panic=abort` for that reason.

Reply:
`panic=abort` seems nice, but to _really_ get it right you need the stdlib to be rebuilt with it. I couldn't get that to work on Alpine, at the very least, though I haven't tried the latest Alpine release.